presto-thrift-connector-api/src/main/java/com/facebook/presto/connector/thrift/api/PrestoThriftPageResult.java

```diff
@@ -120,6 +125,21 @@ public String toString()
                 .toString();
     }
 
+    public static PrestoThriftPageResult fromRecordSet(RecordSet recordSet)
+    {
+        List<Type> types = recordSet.getColumnTypes();
+        int numberOfColumns = types.size();
+        int positions = totalRecords(recordSet);
+        if (numberOfColumns == 0) {
+            return new PrestoThriftPageResult(ImmutableList.of(), positions, null);
+        }
+        List<PrestoThriftBlock> thriftBlocks = new ArrayList<>(numberOfColumns);
+        for (int columnIndex = 0; columnIndex < numberOfColumns; columnIndex++) {
+            thriftBlocks.add(fromRecordSetColumn(recordSet, columnIndex, positions));
```

Comment:
verify that processing the record set multiple times is not "expensive"

Reply:
As far as I see, RecordSet for index join is backed by filtered pages with some transformations applied on top. It should be fine overall.